Feature: ASM anonymous cart binding in Spartacus

cmanciero opened the pull request:
https://jira.tools.sap/browse/CXSPA-153

cmanciero committed:
add assign to customer

cmanciero committed:
update process to use facade

cmanciero committed:
update reference to model

cmanciero committed:
update public api exports

cmanciero committed:
update directory reference

cmanciero committed:
update libraries

cmanciero committed:
updated dependencies

cmanciero committed:
code updates

giancorderoortiz committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
restore saved carts

cmanciero committed:
update method used for bindcart reload

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
remove unused code

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
remove unneeded code

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
remove unused code

justinlee01 commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.html:
```diff
@@ -1,11 +1,58 @@
+    >
+      <input
+        autocomplete="off"
+        placeholder=""
```
No placeholder?

cmanciero replied:
The UI design had no placeholder text

plabadie replied:
Then could we remove the placeholder attribute?

justinlee01 commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.html:
```diff
@@ -1,11 +1,58 @@
+      >
+        <cx-icon
+          class="cx-icon fas fa-times-circle"
+          ng-reflect-type="RESET"
```
You should remove these - I can't find this being used anywhere else.
Also, removing them in my local, they render fine.

justinlee01 commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.html:
```diff
@@ -1,11 +1,58 @@
+      <div role="presentation" class="search-icon">
+        <cx-icon
+          class="cx-icon fas fa-search"
+          ng-reflect-type="SEARCH"
```
You should remove these - I can't find this being used anywhere else.

justinlee01 commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.html:
```diff
@@ -1,11 +1,58 @@
+  <button
+    formcontrolname="logoutCustomer"
+    (click)="logoutCustomer()"
+    data-cy="logout"
```
You should get rid of this - I guess a class selector is preferred.

justinlee01 commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.ts:
```diff
@@ -1,20 +1,47 @@
 export class CustomerEmulationComponent implements OnInit, OnDestroy {
   customer: User;
+  cartId: FormControl = new FormControl();
+  iconTypes = ICON_TYPE;
```
I don't see this being used in the HTML file.

justinlee01 reviewed (commented):
One thing I'm not sure about is the style file - it feels like more specific selectors (i.e. class selectors) are preferred over tag selectors. But it's up to SPA's conventions.

plabadie commented on core-libs/setup/package.json:
```diff
@@ -29,4 +29,4 @@
     "access": "public"
   }
-}
+}
```
Please revert this.  Files should end with a newline char.

There seems to be many files updated in this PR that don't look related to the ticket.  Please revert any changes not related to the scope of the ticket.

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
revert package files

cmanciero committed:
code review changes

plabadie commented on integration-libs/cdc/tsconfig.lib.json:
```diff
@@ -11,18 +11,63 @@
     "inlineSources": true,
     "experimentalDecorators": true,
     "importHelpers": true,
-    "strict": true,
```
This file does not look relevant to the ticket scope.

cmanciero committed:
revert file

plabadie commented on feature-libs/asm/tsconfig.lib.json:
```diff
@@ -8,15 +8,19 @@
+      "@spartacus/checkout": ["dist/checkout"],
+      "@spartacus/checkout/occ": ["dist/checkout/occ"],
+      "@spartacus/checkout/root": ["dist/checkout/root"],
+      "@spartacus/core": ["dist/core"]
```
What is the purpose of the changes in this file?

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/asm/asm.e2e-spec.ts:
```diff
@@ -58,7 +59,7 @@ context('Assisted Service Module', () => {
 
       cy.log('--> Stop customer emulation');
-      cy.get('cx-customer-emulation button').click();
+      cy.get('cx-customer-emulation [data-cy=logout]').click();
```
As a convention, we shouldn't add attributes in the dom like `data-cy=logout` for the sole purpose of tests.
DOM elements retrieved with css selectors.

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/asm/asm.e2e-spec.ts:
```diff
@@ -71,7 +72,65 @@ context('Assisted Service Module', () => {
         '--> Stop customer emulation using the end session button in the ASM UI'
       );
-      cy.get('cx-customer-emulation button').click();
+      cy.get('cx-customer-emulation [data-cy=logout]').click();
```
As a convention, we shouldn't add attributes in the dom like data-cy=logout for the sole purpose of tests.
DOM elements retrieved with css selectors.

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/asm/asm.e2e-spec.ts:
```diff
@@ -71,7 +72,65 @@ context('Assisted Service Module', () => {
+        .get('h2.cx-total')
+        .then(($cartId) => {
+          const text = $cartId.text();
+          assignedCartId = text.replace('Cart #', '').trim();
```
Perhaps we can retrieve the active cart id from the local storage.  It would be more robust than parsing the page's text

plabadie commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.ts:
```diff
@@ -1,20 +1,43 @@
-import { Component, OnDestroy, OnInit } from '@angular/core';
-import { User, UserService } from '@spartacus/core';
+import {
```
I think the cart binding feature should be in its own component.  Adding all this logic to the existing customer emulation components gives too much responsibility to that component.  Smaller more focused components are better to make things easier to maintain and evolve.  It also makes the unit tests more manageable.

plabadie commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.ts:
```diff
@@ -1,20 +1,43 @@
+  cartIdExists: boolean;
   protected subscription = new Subscription();
 
+  @Output()
+  submitEvent = new EventEmitter<{ customerId?: string }>();
```
I don't think this event is used anymore.

plabadie commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.ts:
```diff
@@ -25,12 +48,62 @@ export class CustomerEmulationComponent implements OnInit, OnDestroy {
     this.isCustomerEmulationSessionInProgress$ =
       this.asmComponentService.isCustomerEmulationSessionInProgress();
+
+    this.subscription.add(
+      this.cartId.valueChanges.subscribe((response: string) => {
+        this.cartIdExists = response.trim().length > 0;
+      })
+    );
```
Form field validation should use angular reactive form built-in validation system instead of creating an observable stream for that purpose.

plabadie commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.ts:
```diff
@@ -25,12 +48,62 @@ export class CustomerEmulationComponent implements OnInit, OnDestroy {
+          this.cartId.setValue(response);
+        }
+      })
+    );
```
This logic should be removed.  I see no purpose to init the form field with a value that is never applicable and 100% sure to yield an error.

cmanciero replied:
This logic is here to display the selected customer's active cart id value, if available. This request came from the PO. How would it 100% yield an error, when none were displayed or found during testing?

plabadie replied:
An asm agent, via the bind cart feature, can only bind a cart that does not belong yet to any registered customer (anonymous carts).   The active cart, by definition, is already bound to the current customer.

There is no valid use case, afaik, where the active cart id could be used there.

If the agent clicks to bind that pre-populated cart number, the call will generate an error.  The payload of the error is the following:
```
{
   "errors" : [ {
      "message" : "Only anonymous cart can be bound to customer.",
      "type" : "BadRequestError"
   } ]
}
```

cmanciero replied:
The use case is that the Product Owner would like to display the active cart in the text box so that the Agent knows there is currently a cart assigned to them. I just added functionality that if the Cart id entered or loaded is the previous assigned cart, then the Assign cart to customer button is disabled to prevent an Agent from trying to assign to a customer

plabadie replied:
The active cart id is already displayed on the cart page should the agent want to see it.
Moreover, it is confusing for the agent to see a number in the cart binding input field that:
- he don't know what this cart number is or means
- doesn't understand why he can't submit what appears to be a default value/suggestion.

plabadie replied:
After some discussions: the feature is meant to replicate the ACC behaviour, so it stays.

plabadie commented on feature-libs/asm/package.json:
```diff
@@ -40,4 +42,4 @@
   },
   "schematics": "./schematics/collection.json"
-}
+}
```
This might fail with the prettier formatting or linting validations. Files should end with a new line.

plabadie commented on feature-libs/asm/root/facade/asm-facade.service.ts:
```diff
@@ -0,0 +1,18 @@
+      methods: ['bindCart'],
+    }),
+})
+export abstract class AsmFacadeService {
```
To have a consistent naming, the `AsmFacadeService` should be called `AsmFacade` instead.

Also update the file name.

plabadie commented on feature-libs/asm/core/facade/facade-providers.ts:
```diff
@@ -0,0 +1,11 @@
+import { AsmQueryService } from './asm-query.service';
+
+export const facadeProviders: Provider[] = [
+  AsmQueryService,
+  {
+    provide: AsmFacadeService,
+    useExisting: AsmQueryService,
+  },
```
The Asm facade should be implemented by the existing AsmService. 

I don't see an added value to create a separate service `AsmQueryService`.  Maybe there is.  Let's discuss.

cmanciero replied:
I coordinated with Hak and Eric on their solution with Customer List and we decided this is the best course because 
Putting them together would require the ngrx methods to work with the facade proxy, which we're not sure would be side-effect free.

plabadie replied:
There are ngrx based functions used by most of the facade services.  This is something widespread in the code base.
We need to provide users with a consistent API and I see no need to create a new Facade service to host the new method.  

I was rather thinking about adding the new method in the existing Asm Service and have a facade provider like this:

```
  AsmService,
  {
    provide: AsmFacade,
    useExisting: AsmService,
  },
```

Let's discuss please.

cmanciero replied:
spoke with Eric and Hak, updated code to use AsmService instead, will push shortly

plabadie replied:
The refactoring seems to have evolved into creating a separate service and Facade pair AsmBindCartService and AsmBindCartFacade.  That's the thought process?

plabadie replied:
The refactoring resulted by the creation of BindCartService and BindCartFacade instead.  What is the motivation behind the design?  Was there perhaps too much impact to modify AsmService while avoiding breaking changes?

ercultimate replied:
We were unable to convert AsmService to a facade without inducing breaking changes elsewhere in the module to support the shift to a facade.  Additionally, when trying to use the current new facade as a regular dependency in AsmService, there were injector issues that prevented the OccAdapter from being injected at runtime.  

Creating this granular facade allowed us to begin introducing this design paradigm to the ASM module with the eventual goal of migrating AsmService to a facade in the 6.0 release, and merging any granular facades at that time.

plabadie commented on feature-libs/asm/occ/adapters/occ-asm.adapter.ts:
```diff
@@ -65,4 +66,27 @@ export class OccAsmAdapter implements AsmAdapter {
       .pipe(this.converterService.pipeable(CUSTOMER_SEARCH_PAGE_NORMALIZER));
   }
+
+  bindCart({ cartId, customerId }: BindCartOptions): Observable<void> {
```
Return type should probably be `Observable<unknown>`  to be consistent with the rest

plabadie commented on feature-libs/asm/root/model/bind-cart.model.ts:
```diff
@@ -0,0 +1,4 @@
+export interface BindCartOptions {
```
These are more parameters than options in this case.  Should we name the interface `BindCartParams` `BindCartParameters`?  What do you think?

plabadie commented on feature-libs/asm/root/model/bind-cart.model.ts:
```diff
@@ -0,0 +1,4 @@
+export interface BindCartOptions {
```
Probably this model should be added to the existing model file.

cmanciero replied:
What existing model file? This solution seems to follow the already existing structure.

plabadie replied:
The existing model file I referred to is `feature-libs/asm/core/models/asm.models.ts`

However, the models should really be in the root entry point.  You can leave your model here and instead we should bring over the other ones at some point.

plabadie commented on feature-libs/asm/styles/components/_asm-main-ui.component.scss:
```diff
@@ -19,7 +19,7 @@
 
   .logout {
-    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16' viewBox='0 0 16 16'%3E%3Cpath fill='%23d1e3ff' d='M11,2.7c1.2,0.6,2.2,1.5,2.9,2.6c1.3,1.9,1.5,4.4,0.6,6.5c-0.3,0.8-0.8,1.6-1.5,2.2c-0.6,0.6-1.4,1.1-2.2,1.5 C9.9,15.8,9,16,8,16c-0.9,0-1.9-0.2-2.7-0.5c-0.8-0.4-1.6-0.9-2.2-1.5c-0.6-0.6-1.1-1.4-1.5-2.2C0.7,9.6,0.9,7.2,2.1,5.3 c0.7-1.1,1.7-2, ...
+    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16' viewBox='0 0 16 16'%3E%3Cpath fill='%230064D9' d='M11,2.7c1.2,0.6,2.2,1.5,2.9,2.6c1.3,1.9,1.5,4.4,0.6,6.5c-0.3,0.8-0.8,1.6-1.5,2.2c-0.6,0.6-1.4,1.1-2.2,1.5 C9.9,15.8,9,16,8,16c-0.9,0-1.9-0.2-2.7-0.5c-0.8-0.4-1.6-0.9-2.2-1.5c-0.6-0.6-1.1-1.4-1.5-2.2C0.7,9.6,0.9,7.2,2.1,5.3 c0.7-1.1,1.7-2, ...
```
All the changes to the existing styles will have to be documented as breaking changes.

cmanciero replied:
Done

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/asm/asm.e2e-spec.ts:
```diff
@@ -71,7 +72,65 @@ context('Assisted Service Module', () => {
+      asm.agentSignOut();
+
+      cy.get('button[title="Close ASM"]').click();
+      cy.get('cx-asm-main-ui').should('exist');
+      cy.get('cx-asm-main-ui').should('not.be.visible');
```
I think these assertions are redundant with another test.

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/asm/asm.e2e-spec.ts:
```diff
@@ -71,7 +72,65 @@ context('Assisted Service Module', () => {
+      cy.log(
+        '--> Stop customer emulation using the end session button in the ASM UI'
+      );
+      cy.get('cx-customer-emulation [data-cy=logout]').click();
```
Potentially just asm.agentSignOut(); is enough in the end (lines 130-145). Let's discuss.

plabadie replied:
I think some of the assertions in the end are redundant with the previous test.

plabadie commented on feature-libs/asm/styles/components/_asm-main-ui.component.scss:
```diff
@@ -19,7 +19,7 @@
 
   .logout {
-    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16' viewBox='0 0 16 16'%3E%3Cpath fill='%23d1e3ff' d='M11,2.7c1.2,0.6,2.2,1.5,2.9,2.6c1.3,1.9,1.5,4.4,0.6,6.5c-0.3,0.8-0.8,1.6-1.5,2.2c-0.6,0.6-1.4,1.1-2.2,1.5 C9.9,15.8,9,16,8,16c-0.9,0-1.9-0.2-2.7-0.5c-0.8-0.4-1.6-0.9-2.2-1.5c-0.6-0.6-1.1-1.4-1.5-2.2C0.7,9.6,0.9,7.2,2.1,5.3 c0.7-1.1,1.7-2, ...
+    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16' viewBox='0 0 16 16'%3E%3Cpath fill='%230064D9' d='M11,2.7c1.2,0.6,2.2,1.5,2.9,2.6c1.3,1.9,1.5,4.4,0.6,6.5c-0.3,0.8-0.8,1.6-1.5,2.2c-0.6,0.6-1.4,1.1-2.2,1.5 C9.9,15.8,9,16,8,16c-0.9,0-1.9-0.2-2.7-0.5c-0.8-0.4-1.6-0.9-2.2-1.5c-0.6-0.6-1.1-1.4-1.5-2.2C0.7,9.6,0.9,7.2,2.1,5.3 c0.7-1.1,1.7-2, ...
```
Is updating the styling of the existing interface still in scope of this ticket?

cmanciero replied:
yes

plabadie commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.spec.ts:
```diff
@@ -74,11 +114,67 @@ describe('CustomerEmulationComponent', () => {
     expect(asmComponentService.logoutCustomer).toHaveBeenCalled();
   });
+
+  it('should assign cart to customer', () => {
```
At fist glance, the unit test seems to be written like an e2e test. Unit tests, on the other hand, assert many individual behaviours. (like the component doesn't submit the form if the field is empty.).  Can the test be re-designed?

plabadie reviewed (changes requested):
See comments on the PR for requested changes.

cmanciero committed:
code review request changes

cmanciero committed:
moved asm-bind-cart to own component

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
add type

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
add to index

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
deps generated

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
config:update tsconfig.lib

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
update unit test

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
update e2e test

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
documented style changes

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
added 6.0 changes md

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

plabadie commented on feature-libs/asm/asm.module.ts:
```diff
@@ -2,7 +2,6 @@ import { NgModule } from '@angular/core';
 import { AsmComponentsModule } from '@spartacus/asm/components';
 import { AsmCoreModule } from '@spartacus/asm/core';
 import { AsmOccModule } from '@spartacus/asm/occ';
-
```
Can you revert this to reduce noise in the PR.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,83 @@
+    Validators.required,
+    Validators.minLength(1),
+  ]);
+  isCustomerEmulationSessionInProgress$: Observable<boolean>;
```
`isCustomerEmulationSessionInProgress$` doesn't seem to be used.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,83 @@
+  templateUrl: './asm-bind-cart.component.html',
+})
+export class AsmBindCartComponent implements OnInit, OnDestroy {
+  @Input() customer: User;
+  @Input() cartId: FormControl = new FormControl('', [
```
The current user can be obtained via the API.

plabadie replied:
The form control is not something we want to be provided by a `@Input()`

The form control also needs to live in a form.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,83 @@
+  templateUrl: './asm-bind-cart.component.html',
+})
+export class AsmBindCartComponent implements OnInit, OnDestroy {
+  @Input() customer: User;
```
The current user can be obtained via the API, in which case we don't need a `@Input()`

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,83 @@
+  ) {}
+
+  ngOnInit(): void {
+    this.subscription.add(
```
See other comment and discussion thread about setting the active cart id in the input field.

WeizhengSap replied:
I did not attend the discussion. But, only anonymous cart can be bound to customer. So, why we set user's active cart here?

plabadie replied:
@WeizhengSap you are right, only an anonymous cart can be bound to a customer.  Moreover, the active cart ID is already available on the cart page if ever the agent needs to see it.

The discussion I referred to is [here](https://github.com/SAP/spartacus/pull/15776#discussion_r893863621) on the same PR.
In the end, it appears that this was the behaviour of the ACC UI and the PO decision is to port the behaviour over to Spartacus to achieve feature parity.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,83 @@
+  /**
+   * Assign the input cart number to the customer
+   */
+  assignCartToCustomer() {
```
Suggestion: In some places, the action is call "assign" cart ( like here ans the i18n keys) and in some other places it's called "bind" cart.  To improve readability, perhaps we can consider a consistent naming.

cmanciero replied:
updated to bind*

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,38 @@
+  </div>
+  <button
+    class="assignCartToCustomer"
+    [disabled]="cartId.invalid"
```
It's still possible to submit an empty value by pressing the "enter" key.

How do we intend to deal with this?  Should there be a form validation message or should this be also blocked when the form isn't valid?

cmanciero replied:
If disabled the submit process is also disabled. Even if you remove the disabled attribute from the button, the bind cart method checks for values before submitting

plabadie replied:
If I press enter in the empty/invalid field, the request is sent to the backend and the only message displayed is the one from the backend that said the cart failed to be assigned..

cmanciero replied:
The latest code, not submitted yet, prevents this from happening

ercultimate replied:
In-progress check added to component class.  Please resolve if sufficient.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,38 @@
+<label for="cartNumber"
```
The input field should be in a form.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,38 @@
+    [class.active]="cartId.valid"
+    (click)="cartIdElement.focus()"
+  >
+    <input
```
Probably the input field should be in a form.

plabadie commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.ts:
```diff
@@ -2,7 +2,6 @@ import { Component, OnDestroy, OnInit } from '@angular/core';
 import { Observable, Subscription } from 'rxjs';
 import { AsmComponentService } from '../services/asm-component.service';
-
 @Component({
```
Can you please revert the change to lower the noise in the PR and have a more consistent code formating.

cmanciero committed:
update files to use AsmFacade

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,38 @@
+<label for="cartNumber"
```
The UI seems to need an update to support mobile/narrower screens properly.  To repro, use the mobile view in chrome:
<img width="391" alt="Screen Shot 2022-06-22 at 8 25 19 AM" src="https://user-images.githubusercontent.com/7065346/175030349-2263312d-8cef-496c-b44d-a28dc7aa518e.png">

plabadie replied:
I see some updates to the display, but the UI seems to still need some improvements to it's responsive behaviour.
![Screen Shot 2022-06-28 at 4 59 04 PM](https://user-images.githubusercontent.com/7065346/176287034-c7b4feab-7d05-428d-9bdc-fe9152e1344a.png)
![Screen Shot 2022-06-28 at 4 58 54 PM](https://user-images.githubusercontent.com/7065346/176287037-f32e4eab-995f-4a81-b07c-fb544d2b17a1.png)

plabadie replied:
There is an issue with style versioning in the storefrontapp. (see https://jira.tools.sap/browse/CXSPA-1358)  We should wait until it is fixed before making more assessments about the state of the UI for anon cart binding.

plabadie replied:
I can see the new UI now.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,38 @@
+      <cx-icon class="cx-icon fas fa-search"></cx-icon>
+    </div>
+  </div>
+  <button
```
When the network is slower or the backend takes some time to process the request, there is no work in progress feedback for the user.  This should be handled in some way.

cmanciero replied:
Is there a Spartacus loading message component to use to display that something is happening?

plabadie replied:
The most common UI mechanism to provide quick feedback are spinners that show that processing is happening.
You can look at the ASM customer search, it has a three dot style spinner.
Most of the storefront UI uses a circular spinner.  

The other purpose for the form to have processing state is to prevent multiple submits of the same data.  The form should not be submittable the the post is processing.

ercultimate replied:
Loader added.  Please resolve if sufficient.

cmanciero committed:
code review changes

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
update import location

cmanciero committed:
update use of asmFacade

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
add implements

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
remove root injection

cmanciero committed:
fixed circular dependency issues

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
update service to facade

cmanciero committed:
update unit test

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
update to use asmFacade

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
update asmFacade use

cmanciero committed:
update missed file

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
disable bind cart part for previous assign carts

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

morganm58 commented on feature-libs/asm/styles/components/_asm-main-ui.component.scss:
```diff
@@ -77,14 +77,16 @@
   .asm-bar {
-    color: #fff;
-    background-color: #354a5f;
+    background-color: #ffffff;
```
colors should be variables. This should be 
`background-color: var(--cx-color-inverse);`

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

morganm58 commented on feature-libs/asm/styles/components/_asm-bind-cart.component.scss:
```diff
@@ -0,0 +1,116 @@
+    border-radius: 4px;
+    background-color: #eff1f2;
+    box-shadow: 0px 4px 4px rgb(0 0 0 / 25%);
+    margin-inline-start: 15px;
```
The margin in the visual is 16px between the cart number label and the search box. 
Maybe change the margin-inline-start to `margin-inline-start: 6px`

morganm58 commented on feature-libs/asm/styles/components/_customer-emulation.component.scss:
```diff
@@ -18,23 +20,36 @@
-      background: url("data:image/svg+xml,%0A%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='16' viewBox='0 0 16 16'%3E%3Cpath fill='white' d='M14.8,8c0-3.7-3-6.8-6.8-6.8S1.3,4.2,1.3,8s3,6.8,6.8,6.8S14.8,11.7,14.8,8z M2.6,8c0-3,2.4-5.5,5.5-5.5S13.5,5,13.5,8 s-2.4,5.5-5.5,5.5S2.6,11,2.6,8z M10.7,5.8v4.4c0,0.2-0.2,0.4-0.4,0.4H5.9c-0.2,0-0.4-0.2-0.4-0.4V5.8c0-0.2,0.2-0.4,0.4-0.4h4.4 C10 ...
-        no-repeat 10px center #bb0000;
+  label {
+    margin: 0 15px 0 0;
```
The margin in the visual is 16px between the cart number label and the search box.
this should be `margin-inline-end` so it can work with a directionality rtl ltr.
Maybe change the  `margin-inline-end: 10px;` that way the combine margin will be 16px.

```
margin-inline-end: 10px;
margin-inline-start: 0;
margin-top: 0;
margin-bottom: 0;
```

morganm58 commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,38 @@
+    >
+      <cx-icon class="cx-icon fas fa-times-circle"></cx-icon>
+    </button>
+    <div role="presentation" class="search-icon">
```
when creating new classes we should use the `cx-` naming convention. `class="search-icon"` should be `class="cx-search-icon"` or `class="cx-asm-search-icon"`

morganm58 commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.html:
```diff
@@ -1,11 +1,10 @@
-    placeholder="{{ customer?.name }}, {{ customer?.uid }}"
-  />
-  <button (click)="logoutCustomer()">
+  <div class="customerInfo">
```
`class="customerInfo"` should be `class="cx-customerInfo"` or `class="cx-asm-customerInfo"`

morganm58 commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,38 @@
+      aria-label="Enter cart id"
+    />
+    <button
+      class="reset"
```
`class="reset" `should be `class="cx-reset"` or `class="cx-asm-reset"`

morganm58 commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,38 @@
+  <div
+    role="search"
+    aria-label="Assign a cart id to customer"
+    class="assignCart"
```
when creating new classes we should use the cx- naming convention. 
`class="assignCart"` should be `class="cx-assignCart"`or  `class="cx-asm-assignCart"`

morganm58 commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.html:
```diff
@@ -1,11 +1,10 @@
-  />
-  <button (click)="logoutCustomer()">
+  <div class="customerInfo">
+    <label class="name">{{ customer?.name }}</label>
```
`class="name" should be class="cx-name"` or class="cx-asm-name"`

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.spec.ts:
```diff
@@ -0,0 +1,136 @@
+
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      imports: [
+        StoreModule.forRoot({}),
+        StoreModule.forFeature(ASM_FEATURE, fromReducers.getReducers()),
+      ],
+      declarations: [AsmBindCartComponent, MockTranslatePipe],
```
If all the services/dependencies are mocked, you should not need to inject the store in this test.

cmanciero replied:
Unit test fails if removed

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,92 @@
+    protected asmFacade: AsmFacade,
+    protected activeCartFacade: ActiveCartFacade,
+    protected multiCartFacade: MultiCartFacade,
+    protected userService: UserAccountFacade
```
perhaps align the var name with the type.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.spec.ts:
```diff
@@ -0,0 +1,136 @@
+    userService = TestBed.inject(UserAccountFacade);
+  });
+
+  it('should assign cart to customer', () => {
```
The test seems built like an e2e and should probably be broken down in many unit tests.

cmanciero replied:
broke up the unit tests

plabadie commented on feature-libs/asm/components/asm-toggle-ui/asm-toggle-ui.component.spec.ts:
```diff
@@ -21,23 +24,31 @@ const mockAsmUi: AsmUi = {
     waitForAsync(() => {
       TestBed.configureTestingModule({
-        imports: [I18nTestingModule],
+        imports: [
+          I18nTestingModule,
+          StoreModule.forRoot({}),
+          StoreModule.forFeature(ASM_FEATURE, fromReducers.getReducers()),
+        ],
         declarations: [AsmToggleUiComponent],
```
You shouldn't need to add the store in this test

cmanciero replied:
Unit test fails if removed

plabadie replied:
Seems this change was reverted

plabadie commented on feature-libs/asm/components/customer-selection/customer-selection.component.spec.ts:
```diff
@@ -71,7 +72,7 @@ const MockAsmConfig: AsmConfig = {
   let component: CustomerSelectionComponent;
   let fixture: ComponentFixture<CustomerSelectionComponent>;
-  let asmService: AsmService;
+  let asmService: AsmFacade;
```
we could change the var name for improved readability

plabadie commented on feature-libs/asm/components/customer-selection/customer-selection.component.ts:
```diff
@@ -35,7 +36,7 @@ export class CustomerSelectionComponent implements OnInit, OnDestroy {
   constructor(
     protected fb: FormBuilder,
-    protected asmService: AsmService,
+    protected asmService: AsmFacade,
```
To be consistent with similar changes to other classe in this PR, perhaps we should change the var name as well.

plabadie commented on feature-libs/asm/root/model/bind-cart.model.ts:
```diff
@@ -0,0 +1,4 @@
+export interface BindCartParams {
+  cartId: string;
+  customerId: string;
+}
```
Might as well move this to asm.models.ts

plabadie commented on feature-libs/asm/core/store/selectors/customer-search.selectors.spec.ts:
```diff
@@ -27,7 +27,7 @@ describe('Customer Search Results Selectors', () => {
       imports: [
         StoreModule.forRoot({}),
-        StoreModule.forFeature('asm', fromReducers.getReducers()),
+        StoreModule.forFeature(ASM_FEATURE, fromReducers.getReducers()),
```
The ASM_FEATURE was intended for lazy loading mechanism and not for the state slice.  It's probably safer not to use it to to identify state slices.

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
code review cleanup

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

morganm58 commented on feature-libs/asm/styles/components/_customer-selection.component.scss:
```diff
@@ -13,9 +13,40 @@
+          display: flex;
+          outline: 0;
+          border: 1px solid #89919a;
+          color: #fff;
```
I'm not sure if asm uses color variables. But if your using a hex value it should be the full version. #ffffff

morganm58 reviewed (changes requested):
I left a few comments on some pieces that should be changed.

morganm58 commented:
I'm not sure if asm is styled for mobile users. If it is the mobile viewport has a few issues.
<img width="890" alt="Screen Shot 2022-06-29 at 1 30 41 PM" src="https://user-images.githubusercontent.com/2947470/176499750-c4f441f4-def4-4561-855c-0e558bc984d9.png">
<img width="670" alt="Screen Shot 2022-06-29 at 1 30 50 PM" src="https://user-images.githubusercontent.com/2947470/176499758-6beca289-cf36-4c2a-8929-f97e3ebda69b.png">

cmanciero committed:
style changes

cmanciero committed:
styl eupdates

cmanciero committed:
add active style back

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

morganm58 commented on feature-libs/asm/styles/components/_asm-bind-cart.component.scss:
```diff
@@ -0,0 +1,99 @@
+    box-shadow: 0px 4px 4px rgb(0 0 0 / 25%);
+    margin-inline-start: 6px;
+    margin-inline-end: 15px;
+    // min-width: 20rem;
```
If this isn't being used it should be removed

cmanciero committed:
remove commented out code

cmanciero committed:
fixunit tests

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
feature flag style and components

cmanciero committed:
Merge branch 'feature/CXSPA-153-asm' of https://github.com/SAP/spartacus into feature/CXSPA-153-asm

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
dep update

cmanciero committed:
update unit test

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

cmanciero committed:
Merge branch 'develop' into feature/CXSPA-153-asm

developpeurweb commented:
The suggest drop-down list must be keyboard accessible. (E.g. Spartacus Main search box or Quick Order search box)
At the moment, it can only be accessed with the mouse.

<img width="1083" alt="Screen Shot 2022-07-12 at 5 08 18 PM" src="https://user-images.githubusercontent.com/15113219/178742985-413c20a1-25ab-4ee2-bfb5-da5d77afc63b.png">
.

ercultimate committed:
add loader to bind-cart button while request is in progress

ercultimate committed:
remove unnecessary store dependency from unit tests

kimhw0630 committed:
Merge branch 'develop' into feature/CXSPA-153-asm

WeizhengSap commented on feature-libs/asm/occ/adapters/occ-asm.adapter.spec.ts:
```diff
@@ -6,10 +6,9 @@ import {
-  CustomerSearchPage,
   CUSTOMER_SEARCH_PAGE_NORMALIZER,
 } from '@spartacus/asm/core';
+import { CustomerSearchOptions, CustomerSearchPage } from '@spartacus/asm/root';
```
Please add unit test for the new function `bindCart`

WeizhengSap commented on feature-libs/asm/root/index.ts:
```diff
@@ -0,0 +1,2 @@
+export * from './facade/index';
+export * from './model/index';
```
I don't think we need this file. Or, we can put them in `index.ts`
```
export * from './asm-constants';
export * from './asm-loader.module';
export * from './asm-root.module';
export * from './facade/index';
export * from './feature-name';
export * from './model/index';
export * from './services/index';
```
and `export * from './index';` in `public_api.ts`.

ercultimate replied:
Exported `asm/root/index.ts` in `public_api.ts`

WeizhengSap commented on feature-libs/asm/core/connectors/asm.adapter.ts:
```diff
@@ -11,4 +12,6 @@ export abstract class AsmAdapter {
     options: CustomerSearchOptions
   ): Observable<CustomerSearchPage>;
+
+  abstract bindCart(options: BindCartParams): Observable<unknown>;
```
Can we add some doc comment?

WeizhengSap commented on feature-libs/asm/core/facade/asm.service.spec.ts:
```diff
@@ -20,25 +22,47 @@ const mockUser: User = {
+  ): Observable<CustomerSearchPage> {
+    return of();
+  }
+  bindCart(_cartId: string, _customerId: string): Observable<unknown> {
```
please add unit test for the new function `bindCart`.

justinlee01 replied:
See `feature-libs/asm/occ/adapters/occ-asm.adapter.ts`.

WeizhengSap commented on feature-libs/asm/components/asm-components.module.ts:
```diff
@@ -22,6 +32,12 @@ import { defaultAsmLayoutConfig } from './default-asm-layout.config';
+    IconModule,
+    PopoverModule,
+    ModalModule,
+    NgSelectModule,
```
`PopoverModule`, `ModalModule`, `NgSelectModule`, `PasswordVisibilityToggleModule` are not used, right? Please check them.

WeizhengSap commented on feature-libs/asm/assets/translations/en/asm.ts:
```diff
@@ -26,6 +26,12 @@ export const asm = {
+      cartNumber: 'Cart Number',
+      bindCartToCustomer: 'Assign Cart to Customer',
+      success: 'Cart has been successfully assigned',
+      error: 'Cart failed to be assigned',
```
this error message is generic. Can we catch the error and display the error message sent from backend?

WeizhengSap commented:
The button of "Close Asm" is invisible.
![Screen Shot 2022-08-02 at 4 48 44 PM](https://user-images.githubusercontent.com/44440575/182644863-f2e34d1b-5c71-41c5-9bc5-9ad4e12f2e38.png)

WeizhengSap commented on feature-libs/asm/occ/adapters/occ-asm.adapter.ts:
```diff
@@ -65,4 +68,27 @@ export class OccAsmAdapter implements AsmAdapter {
+      }
+    );
+
+    return this.http.post<void>(url, {}, { headers, params });
```
please consider to have this `catchError((error) => throwError(normalizeHttpError(error)))`, so the caller can get the error. "Cart failed to be assigned" is too generic.

WeizhengSap commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,101 @@
+          });
+        })
+          .pipe(finalize(() => (this.loading = false)))
+          .subscribe(
```
since we have `subscribe` here, why we use `defer`? can we just do `this.asmFacade.bindCart(...).subscribe(...)?`.

WeizhengSap commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,101 @@
+  templateUrl: './asm-bind-cart.component.html',
+})
+export class AsmBindCartComponent implements OnInit, OnDestroy {
+  @Input() customer: User;
```
why use `@Input()`? the customer is not passed from parent component.

WeizhengSap commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,101 @@
+
+  ngOnInit(): void {
+    this.subscription.add(
+      this.userAccountFacade.get().subscribe((user) => {
```
`this.userAccountFacade.get()` can be moved to command. We can make `customerId` in `BindCartParams` as optional. So, only pass `cartId` to `bindCart` function. In command, something like this:
```
protected bindCartCommand$: Command<BindCartParams> =
    this.commandService.create((options: BindCartParams) =>
      options.customerId
        ? this.asmConnector.bindCart(options)
        : this.userAccountFacade.get().pipe(
            filter(isNotUndefined),
            map((user) => user.uid ?? ''),
            switchMap((customerId) =>
              this.asmConnector.bindCart({
                customerId,
                cartId: options.cartId,
              })
            )
          )
    );
```

ercultimate replied:
The use of `UserAccountFacade` has been encapsulated in the use of a command.  I thought the suggestion was difficult to follow, so I used a different implementation.  Please review and provide any feedback.

plabadie replied:
After review and discussion with @WeizhengSap, here is the feedback:
The parameters passed to the Command is not really the issue here after all.

The intent behind the Command pattern is for it to be used by facade implementations.

The components should be calling a facade method that may be implemented with a command, but as a rule of thumb, components should not create commands directly.  

To improve the design here the component should call a facade method.  In this case, it's ok for the facade method to accept just the cart id and fetch the current user id from within its implementation.  Many facade methods operate on the currently signed in user and retrieve it themselves.  Probably the cart reload should still be triggered by the component upon successfully binding a cart.

Consider the usage of `UserIdService.takeUserId` to retrieve the userId.  This is what many (perhaps most) facade implementations use.  There are many examples in the code base.

WeizhengSap commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.html:
```diff
@@ -0,0 +1,43 @@
+    <span [attr.aria-hidden]="loading">
+      {{ 'asm.bindCart.bindCartToCustomer' | cxTranslate }}
+    </span>
+    <cx-dot-spinner
```
Can we also use this `cx-dot-spinner` in customer selection?

justinlee01 committed:
Removing customer Input from AsmBindCartComponent
No longer using defer in AsmBindCartComponent
Using error message for AsmBindCartComponent instead of using hardcoded string
Removing unnecessary imported modules for AsmComponentsModule
Using DotSpinner for CustomerSelectionComponent
Adding documentation for bindCart method in AsmAdapter contract
Adding unit test for bindCart in OccAsmAdapter
Exchanging contents of index.ts and public_api.ts for asm module

justinlee01 committed:
prettier:fix

justinlee01 committed:
Normalizing error response from OCC bindCart

justinlee01 commented:
@WeizhengSap I have tried my best to respond to your comments; for the others, I will have to wait for my colleagues to come back from vacation who are more familiar with the feature.

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-153-asm

justinlee01 committed:
Fixing ASM E2E tests and linting.

justinlee01 commented:
I have made fixes for the E2E tests. Of particular note is line 36 of `projects/storefrontapp-e2e-cypress/cypress/helpers/applied-promotions.ts`, where I change a `cy.intercept` to `interceptGet` as it seemed more reliable. Let me know what your thoughts are.

justinlee01 committed:
Removing ng-select/ng-select dependency from ASM library

justinlee01 committed:
Converting all relative links to other ASM packages to absolute paths. Also removing ng-select from schematics dependencies

justinlee01 committed:
Changing invalid style in _customer-selection.component.scss

justinlee01 committed:
Fixing hardcoded aria labels in AsmBindCartComponent

justinlee01 committed:
Updating dependency-utils spec as ASM is using cart library, but need to double check if this is appropriate.

justinlee01 committed:
Merge branch 'develop' into feature/CXSPA-153-asm

justinlee01 commented:
Another note. I have updated the expectations of two tests in `projects/schematics/src/shared/utils/dependency-utils_spec.ts` related to CDC - the ASM module now uses the Cart module.

justinlee01 committed:
Rerun tests

hakwookim committed:
updated ASM close button to show

ercultimate committed:
move asm models back to core

ercultimate committed:
migrate facade methods back to service to avoid breaking changes

ercultimate committed:
remove breaking change to spinner html in customer-selection.component

- delete 6.0 change notes

ercultimate committed:
remove unused dependency from asm.service

ercultimate committed:
fix misspelling

ercultimate committed:
Merge remote-tracking branch 'origin/develop' into feature/CXSPA-153-asm

ercultimate committed:
update asm-bind-cart unit test

ercultimate committed:
clairfy asm-bind-cart component unit tests

ercultimate committed:
update asm e2e test

hakwookim committed:
Fix for accessibility problem in search box

hakwookim committed:
Fix for accessibility problems in login screen

hakwookim committed:
Add missing license header to sources

ercultimate committed:
refactor bind cart component to use a command for binding

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.spec.ts:
```diff
@@ -0,0 +1,177 @@
+
+      component.bindCartToCustomer();
+
+      expect(asmBindCartFacade.bindCart).toHaveBeenCalledTimes(0);
```
or ```.not.toHaveBeenCalled();```

plabadie replied:
or `.not.toHaveBeenCalled();`

ercultimate replied:
Is this preferred convention?

plabadie commented on feature-libs/asm/core/facade/asm-bind-cart.service.ts:
```diff
@@ -0,0 +1,28 @@
+import { AsmConnector } from '../connectors';
+
+@Injectable()
+export class AsmBindCartService implements AsmBindCartFacade {
```
This class needs a corresponding unit test.  (even though the methods don't contain alot of logic)

plabadie commented on feature-libs/asm/core/facade/asm-bind-cart.service.ts:
```diff
@@ -0,0 +1,28 @@
+import { AsmConnector } from '../connectors';
+
+@Injectable()
+export class AsmBindCartService implements AsmBindCartFacade {
+  constructor(
```
Ideally, I think the bindCart function could have gone in the AsmService class.  However I understand it's more overhead to make it backwards compatible.  Perhaps there were also additional motivations to create a new distinct service class that I'm not aware of. 

It's your call.

plabadie commented:
The UI seems to be in bad shape still.  The potential issues are numerous:  misaligment, button styling seems missing or broken, icons appear inside buttons where they shouldn't, ui doesn't seem to respond well to smaller widths.  I took  screenshots:

Also please test for a11y.

<img width="538" alt="Screen Shot 2022-09-02 at 3 15 29 PM" src="https://user-images.githubusercontent.com/7065346/188222274-88137dc0-74be-4ed9-a6c3-9ace22b3867a.png">
<img width="537" alt="Screen Shot 2022-09-02 at 3 15 17 PM" src="https://user-images.githubusercontent.com/7065346/188222278-5e4e3dcf-cf04-471b-b5ca-002e3f63e7d4.png">
<img width="582" alt="Screen Shot 2022-09-02 at 3 14 48 PM" src="https://user-images.githubusercontent.com/7065346/188222280-6832004e-6192-4675-9a3f-84bd7e783752.png">
<img width="967" alt="Screen Shot 2022-09-02 at 3 14 04 PM" src="https://user-images.githubusercontent.com/7065346/188222282-99f8d533-e50c-4ffb-8136-d7fb3aff8e21.png">
<img width="966" alt="Screen Shot 2022-09-02 at 3 13 19 PM" src="https://user-images.githubusercontent.com/7065346/188222284-1828b10a-a5f8-4872-a490-147c91972fc8.png">

plabadie commented:
Unless this is intentional, there ia a regression in the agent login form.  We don't see the eye button anymore in the password field.  See screenshots that compare develop and the feature branch:

<img width="1178" alt="Screen Shot 2022-09-02 at 3 18 57 PM" src="https://user-images.githubusercontent.com/7065346/188222827-f2ca6f5e-6863-4891-8b98-f9bd520ad20e.png">
<img width="980" alt="Screen Shot 2022-09-02 at 1 47 01 PM" src="https://user-images.githubusercontent.com/7065346/188223055-e3a8a04f-23bd-4414-a1a0-4b5d755f164a.png">

plabadie commented on feature-libs/asm/assets/translations/en/asm.ts:
```diff
@@ -32,6 +32,16 @@ export const asm = {
+      cartNumber: 'Cart Number',
+      bindCartToCustomer: 'Assign Cart to Customer',
+      success: 'Cart has been successfully assigned',
+      aria: {
```
I don't think we make a distinction for aria labels anywhere.  For consistency, perhaps remove the aria level.

justinlee01 replied:
For the Iceman working on this ticket, this is a change I made.

plabadie replied:
...however I did see around 10 cases of labels with the pattern `ariaLabelSomeName` in the other translation files.  The use cases seem to be when a label had a different content in the aria tag context, so 2 labels were needed.  If this is the same use case perhaps we can use that pattern.

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,118 @@
+
+@Component({
+  selector: 'cx-asm-bind-cart',
+  templateUrl: './asm-bind-cart.component.html',
```
Requires `changeDetection: ChangeDetectionStrategy.OnPush,`

plabadie commented on feature-libs/asm/components/customer-selection/customer-selection.component.spec.ts:
```diff
@@ -129,6 +129,7 @@ describe('CustomerSelectionComponent', () => {
     component.ngOnInit();
     fixture.detectChanges();
 
+    // TODO: (CXSPA-153) Change div.spinner to use cx-dot-spinner component
```
Is that todo still relevant?
Also, beware of breaking changes when changing the UI.

ercultimate replied:
This was originally changed to use `<cx-dot-spinner>` but was reverted for breaking changes.  This todo is for the 6.0 migration.

plabadie replied:
As discussed we can create the follow up ticket for 6.0 already and refer to it in the comment instead of CXSPA-153.

ercultimate replied:
Todo ticket updated to task CXSPA-1291

plabadie commented on feature-libs/asm/components/customer-selection/customer-selection.component.html:
```diff
@@ -5,13 +5,14 @@
     ></cx-form-errors>
   </label>
-  <button type="submit">
+  <button type="submit" [class.active]="selectedCustomer">
```
I suspect this is a breaking change.  We can discuss.

Also, I think the correct way to implement this is to enable/disable the submit button based on the form validity.  This way the validation rules are driven only by the form in the component and all the template needs to know is if the form is valid.

plabadie replied:
I discussed this with others. While we do not yet have a clearly defined rule on how to handle these cases, we agree that we can't guarantee that adding classes won't collide with customer classes/styling.  To unblock things in this PR and to be able to move forward, let's use the feature level directive in the html to add the class in 5.1+.

plabadie commented on feature-libs/asm/components/dot-spinner/dot-spinner.component.ts:
```diff
@@ -0,0 +1,13 @@
+
+@Component({
+  selector: 'cx-dot-spinner',
+  templateUrl: './dot-spinner.component.html',
```
Requires `changeDetection: ChangeDetectionStrategy.OnPush,`

ercultimate replied:
Updated. Please resolve if sufficient.

plabadie commented on feature-libs/asm/components/index.ts:
```diff
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+export * from './asm-bind-cart/asm-bind-cart.component';
```
Should we also export the dot spinner?

Is the dot spinner reusable enough so we add it in core instead of ASM?  Or perhaps it's tailored tightly to the ASM UI.

ercultimate replied:
I think it is expected to be part of the ASM module's style, which is distinct from the Spartacus styles.  It does make sense for the ASM component module to export it.

plabadie commented on feature-libs/asm/occ/adapters/occ-asm.adapter.ts:
```diff
@@ -71,4 +74,29 @@ export class OccAsmAdapter implements AsmAdapter {
       .pipe(this.converterService.pipeable(CUSTOMER_SEARCH_PAGE_NORMALIZER));
   }
+
+  bindCart({ cartId, customerId }: BindCartParams): Observable<unknown> {
```
could we just have `bindCart(options: BindCartParams)`.
Is there an advantage to destructure the params in the declaration?

If ever we add attributes to the BindCartParams interface, I fear we need to then update the method signature, which is problematic for backwards compatibility.

ercultimate replied:
Destructuring is primarily used for code readability.  
The signature for the `bindCart()` method _is_ `bindCart(options: BindCartParams)`.  

It is functionally equivalent to
```
  bindCart(arg0: BindCartParams): Observable<unknown> {
    const cartId = arg0.cartId;
    const customerId = arg0. customerId
```

The destructuring is just additional syntax to extract particular properties of the parameter object into variables, with the intent of better readability.  Adding a new property to `BindCartParams` would require a text change to the method declaration, but would not result in a change to the signature.

plabadie commented on projects/schematics/src/shared/utils/dependency-utils_spec.ts:
```diff
@@ -106,7 +112,13 @@ describe('dependency-util', () => {
       ]);
 
-      expect(result).toEqual([SPARTACUS_USER, SPARTACUS_ASM, SPARTACUS_CDC]);
+      // TODO: Double-check this.
```
Why are schematics unit tests updated in the scope of this PR?
Perhaps clarifying this is the purpose of the todo.

justinlee01 replied:
I added this TODO. I believe the reason for this is because the ASM module now depends on the Cart module, therefore this list of dependencies needs to be updated. The TODO is to check with a senior SPA dev.

Edit: An additional note, this is related to this CI check: "Config check / Dependencies and tsconfig files".

ercultimate replied:
@justinlee01 Can you explain this?

ercultimate replied:
@plabadie I am awaiting your comment.

plabadie commented on projects/schematics/src/shared/utils/dependency-utils_spec.ts:
```diff
@@ -79,7 +79,13 @@ describe('dependency-util', () => {
 
-      expect(result).toEqual([SPARTACUS_USER, SPARTACUS_ASM, SPARTACUS_CDC]);
+      // TODO: Double-check this.
+      expect(result).toEqual([
```
Why are schematics unit tests updated in the scope of this PR?
Perhaps clarifying this is the purpose of the todo.

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/integration/regression/asm/asm.e2e-spec.ts:
```diff
@@ -48,6 +51,62 @@ context('Assisted Service Module', () => {
+        .then(($cartId) => {
+          // localStorage contains anonymous cart uid, read code from UI
+          const text = $cartId.text();
+          anonymousCartCode = text.replace('Cart #', '').trim();
```
Cart id could perhaps also be retrieved from local storage here as well.
Could doing so also avoid to go to the cart page?

ercultimate replied:
The comment on line 72 was intended to explain why this is reading the code from the UI.  The local storage contains the cart uid for an anonymous cart, but an authenticated-user-bound cart the local storage contains the cart code.  We cannot compare them directly.

plabadie commented on projects/storefrontapp/src/app/app.module.ts:
```diff
@@ -84,7 +84,7 @@ if (!environment.production) {
     provideConfig(<FeaturesConfig>{
       features: {
-        level: '4.2',
+        level: '5.1',
```
This soon to be merged PR should resolve that once and for all: https://github.com/SAP/spartacus/pull/16116/files

plabadie commented on projects/storefrontapp-e2e-cypress/cypress/helpers/applied-promotions.ts:
```diff
@@ -38,11 +39,10 @@ export function checkForAppliedPromotions() {
+  interceptGet(
+    'cart_refresh',
+    '/users/*/carts/*?fields=DEFAULT,potentialProductPromotions*'
+  );
```
Why update the promotions e2e in this PR?

justinlee01 replied:
I worked on this part. I noticed the tests break when pointing to a local OCC source (due to some error in calculating API_URL). This is to make the set-up consistent and possibly prevent flakiness.

ercultimate committed:
change translation keys to remove aria level in asm

ercultimate committed:
update change detection to OnPush and adapt componenets to change

ercultimate committed:
restore lost module import and export new component

ercultimate committed:
update unit test with more common assertion form

ercultimate committed:
fix library compilation errors

ercultimate committed:
update todo note with proper ticket number

ercultimate committed:
Merge remote-tracking branch 'origin/develop' into feature/CXSPA-153-asm

ercultimate committed:
refactor asm bind cart to use command in facade layer

ercultimate committed:
add feature level restriction to active class in customer selection button

ercultimate committed:
Merge remote-tracking branch 'origin/develop' into feature/CXSPA-153-asm

ercultimate committed:
update unit tests for asm anonymous cart

ercultimate committed:
remove index.ts in asm/root package

ercultimate committed:
fix import for core style lib in asm component styles

ercultimate committed:
remove unneeded todo comment

ercultimate committed:
Merge remote-tracking branch 'origin/develop' into feature/CXSPA-153-asm

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -86,13 +56,14 @@ export class AsmBindCartComponent implements OnInit, OnDestroy {
-        concatMap(() => this.bindCartToCurrentUser$.execute(this.cartId.value)),
+        concatMap(() => this.asmBindCartFacade.bindCart(this.cartId.value)),
+        tap(() => {
+          this.multiCartFacade.reloadCart(OCC_CART_ID_CURRENT);
```
When we all reloadCart in the tap(), are we sure at that point that the cart binding was a success?

Otherwise, it could also improve readability to group together the logic to be applied when the cart binding is a success.  We'd just need to move the reloadCart before we display the success message display.  There might also be something I missed.  Perhaps there is a reason for how it is implemented like error catching.

ercultimate replied:
Yes, the binding must be a success.  The `tap()`'s "next" listener executes only when  `asmBindCartFacade.bindCart()` emits.  If it "error"s, then the tap is passed by.  If the filter above causes a "complete", the tap would also be passed by.

I agree it would be more readable to move the `tap()`'s content to the `subscribe()`'s "next" listener.

plabadie commented on feature-libs/asm/core/facade/asm-bind-cart.service.ts:
```diff
@@ -5,24 +5,45 @@
+    { cartId: string; userId?: string },
+    unknown
+  > = this.commandService.create(
+    ({ cartId, userId }: { cartId: string; userId?: string }) =>
```
Do we need to accept a userId param?  It seems that what is used by the UI component is the cartId only method signature.  If it is not required, dropping it would simplify the method.

plabadie commented on feature-libs/asm/root/public_api.ts:
```diff
@@ -4,4 +4,10 @@
  */
 
-export * from './index';
+export * from './asm-constants';
```
Good catch removing the index.ts

plabadie committed:
CXSPA-153: Apply temporary fix for style versioning issue.

plabadie committed:
Merge branch 'develop' into feature/CXSPA-153-asm

plabadie commented on feature-libs/asm/components/asm-bind-cart/asm-bind-cart.component.ts:
```diff
@@ -0,0 +1,96 @@
+  templateUrl: './asm-bind-cart.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class AsmBindCartComponent implements OnInit, OnDestroy {
```
Just for the sake of discussion, the use of the ASM for classes and interfaces started back when ASM was not in its own feature library.  It is always case by case, but it is perhaps less useful now and a bit redindant.  I'm not insisting on a change of name for the new classes in the PR.  I just want to point it out in case it allows for leaner naming in the future.

ercultimate replied:
Thanks, it's good to know if something is convention or coincidence. 👍

ercultimate committed:
refactor bind to cart component for readability

ercultimate committed:
remove customer ID parameter form bind cart facade

ercultimate committed:
Merge remote-tracking branch 'origin/develop' into feature/CXSPA-153-asm

plabadie commented on feature-libs/asm/root/facade/asm-bind-cart.facade.ts:
```diff
@@ -0,0 +1,27 @@
+export abstract class AsmBindCartFacade {
+  /**
+   * Bind an anonymous cart to a registered user
+   * @param userId The user to associate a cart with.  Defaults to the current user if not provided.
```
The doc comment needs updating to remove the userId param (and add cartid perhaps)

ercultimate committed:
fix incorrect doc comment on bind cart method

ercultimate committed:
Merge remote-tracking branch 'origin/develop' into feature/CXSPA-153-asm

justinlee01 commented on .gitignore:
```diff
@@ -3,6 +3,7 @@
 # compiled output
 /dist
 /tmp
+/temp
```
What creates this `/temp` directory? Going by the commit, is it from unit tests?

ercultimate replied:
API Extractor

justinlee01 commented on feature-libs/asm/components/customer-emulation/customer-emulation.component.spec.ts:
```diff
@@ -3,7 +3,7 @@ import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
 import { I18nTestingModule, User } from '@spartacus/core';
 import { UserAccountFacade } from '@spartacus/user/account/root';
-
+import { MockFeatureLevelDirective } from 'projects/storefrontlib/shared/test/mock-feature-level-directive';
```
I believe this should be a path to a package, but probably does not hurt the compilation of the tests.

ercultimate replied:
The class is not exported as part of a module.  This is the only valid import path.

justinlee01 reviewed (approved):
Should we create a ticket for 6.0 to remove the 5.1 flags?
I will leave it up to your discretion whether we should host a code review with the team on this. The code seems clear to me, though I have spent some time with the code already.

ercultimate committed:
retry tests with empty commit

ercultimate committed:
Merge branch 'develop' into feature/CXSPA-153-asm
